feat(alert): add ability to use alert inline

amandaejohnston opened the pull request:
<!-- Please refer to our contributing documentation for any questions on submitting a pull request, or let us know here if you need any help: https://ionicframework.com/docs/building/contributing -->

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
  - Some docs updates need to be made in the `ionic-docs` repo, in a separate PR. See the [contributing guide](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#modifying-documentation) for details.
- [ ] Build (`npm run build`) was run locally and any changes were pushed
- [ ] Lint (`npm run lint`) has passed locally and any fixes were made for failures


## Pull request type

<!-- Please do not submit updates to dependencies unless it fixes an issue. --> 

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. --> 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying. -->


<!-- Issues are required for both bug fixes and features. -->
Issue URL: 


## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

amandaejohnston commented on core/src/components/alert/alert.scss:
```diff
@@ -52,6 +52,10 @@
   align-items: flex-start;
 }
 
+:host ion-backdrop {
+  z-index: 0;
+}
```
This was added because I was running into an issue where inline alerts would show the backdrop on top of the overlay. I figure the backdrop's default `z-index` of 2 isn't needed since the overlay wrapper around it gets its own massive z-index anyway.

amandaejohnston commented on core/src/components/alert/alert.scss:
```diff
@@ -54,6 +54,7 @@
 
 .alert-wrapper {
   display: flex;
+  position: relative;
```
This was added to resolve an issue where inline alerts show underneath the backdrop, due to the added `ion-delegate-host` wrapper messing up the stacking.

liamdebeasi replied:
Hmm this looks like a bug. `ion-delegate-host` should only be added to `ion-modal` and `ion-popover` since they accept user components.

sean-perkins replied:
`ion-delegate-host` doesn't have any style responsibility. It was just used to track the existing wrapper element that was created, so we didn't create another one. e.g. During the PR at one point I was setting a private flag vs. adding a class. We landed on a class just for simplicity and to avoid someone accidentally setting that flag. 

I would need more information to what caused this style being introduced.

amandaejohnston commented on core/src/components/alert/test/a11y/alert.e2e.ts:
```diff
@@ -41,7 +41,12 @@ test.describe('alert: a11y', () => {
+     * Axe will be unable to verify color contrast
+     * on elements under the backdrop.
+     */
+    const results = await new AxeBuilder({ page }).disableRules('color-contrast').analyze();
```
Adding `position: relative` to the alert seems to have caused the color contrast test to start failing and/or exacerbated an existing issue, but I can't see any difference visually, so it seems to be an Axe detection issue rather than an actual contrast problem.

liamdebeasi replied:
Since there are layered elements with opacity (such as the backdrop), the `position: relative` with `z-index: 10` likely creates a new stacking context which can cause Axe to be unable to verify the color contract.

liamdebeasi reviewed (approved):
The `position: relative` change is concerning, but it looks like this is due to the `ion-delegate-host` bug. Once that is resolved we should be able to remove `position: relative`. Looks great otherwise!

liamdebeasi reviewed (approved):
Great job!

amandaejohnston commented:
After talking on Slack, I've patched the `ion-delegate-host` element to only be added for modals and popovers, and reverted the patches to the alert CSS and a11y test. I also added a dev build to the bottom of the PR description. 👍
